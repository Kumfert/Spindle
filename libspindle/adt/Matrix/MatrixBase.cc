//
// MatrixBase.h -- a base class for all matrix/graph classes
//
//  $Id: MatrixBase.cc,v 1.2 2000/02/18 01:31:49 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//


#include "spindle/MatrixBase.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

bool 
MatrixBase::resize( const int newNCols, const int newNRows, const int newNNonZeros ) {
  
  colPtr.resize( newNCols + 1 );
  rowIdx.resize( newNNonZeros );
  if ( diagPtr.notNull() ) {
    if ( diagPtr.size() == nCols ) {
      diagPtr.resize( newNCols );
    } else { 
      diagPtr.resize( 0 );
    }
  }

  if ( realValues.notNull() ) {
    if ( realValues.size() == nNonZeros ) {
      realValues.resize( newNNonZeros );
    } else {
      realValues.resize( 0 );
    } 
  }
  
  if ( imgValues.notNull() ) {
    if ( imgValues.size() == nNonZeros ) {
      realValues.resize( newNNonZeros );
    } else {
      realValues.resize( 0 );
    }
  }

  nRows = newNRows;
  nCols = newNCols; 
  nNonZeros = newNNonZeros;
  return true;
}

bool
MatrixBase::setDiagPtr() {
  bool recomputeDiagPtr = false; 
  const int n = nCols;
  if ( diagPtr.size() != n ) {
    // there is no diagPtr, so create it
    diagPtr.resize( n );
    recomputeDiagPtr = true;
  }
  // if we aren't already sure that we must recompute
  if (!recomputeDiagPtr) { // test all values
    for(int i=0; i<n; ++i ) {
      if (rowIdx[diagPtr[i]] != i) {
        recomputeDiagPtr = true;
        break;
      }
    }
  }
  // Now recompute if neccessary
  if (recomputeDiagPtr) {
    const int* colptr = colPtr.lend();
    const int* rowidx = rowIdx.lend();
    for(int i=0; i<n ;i++) {
      for(int j=colptr[i];j<colptr[i+1];j++){
        diagPtr[i]=j;
        if(rowidx[j] == i) {
          break;
        } else if (rowidx[j] > i) {
          // error, should have explicit self edges
          return false;
        }
      }
    }
  }
  return true;
}
