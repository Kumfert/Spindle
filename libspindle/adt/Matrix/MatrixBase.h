//
// MatrixBase.h -- a base class for all matrix/graph classes
//
//  $Id: MatrixBase.h,v 1.2 2000/02/18 01:31:50 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_MATRIX_BASE_H_
#define SPINDLE_MATRIX_BASE_H_

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

SPINDLE_BEGIN_NAMESPACE 

class MatrixUtils;

class MatrixBase { 
friend class MatrixUtils;

protected:
  int nRows;
  int nCols;
  int nNonZeros;

  SharedArray<int> colPtr;
  SharedArray<int> rowIdx;
  SharedArray<int> diagPtr;
  SharedArray<double> realValues;
  SharedArray<double> imgValues;

public:

  bool resize( const int newNCols, const int newNRows, const int newNNonZeros );
  int queryNCols() const { return nCols; }
  int queryNRows() const { return nRows; }
  int queryNNonZeros() const { return nNonZeros; }

  const SharedArray<int>& getColPtr() const { return colPtr;  }
  const SharedArray<int>& getRowIdx() const { return rowIdx;  }
  const SharedArray<int>& getDiagPtr() const { return diagPtr;  }
  const SharedArray<double>& getRealValues() const { return realValues; }
  const SharedArray<double>& getImgValues() const { return imgValues;  }
  
  SharedArray<int>& getColPtr() { return colPtr; }
  SharedArray<int>& getRowIdx() { return rowIdx; }
  SharedArray<int>& getDiagPtr() { return diagPtr; }
  SharedArray<double>& getRealValues() { return realValues; }
  SharedArray<double>& getImgValues() { return imgValues;  }

  bool setDiagPtr();
};

SPINDLE_END_NAMESPACE

#endif
