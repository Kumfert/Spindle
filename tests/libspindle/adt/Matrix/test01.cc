//
// testMatrixBase_1.cc --tests basic MatrixBase operations
//
//
//
//

#include <iostream.h>
#include <iomanip.h>
#include "spindle/MatrixBase.h"
#include "samples/small_mesh.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

//
// This is a simple derived class to set some of the information
// conveniently.  "MatrixBase" wasn't designed to be used directly.
//
class MatrixTest : public MatrixBase { 
public: 
  void setNRows( int nrows ) { nRows = nrows; }
  void setNCols( int ncols ) { nCols = ncols; }
  void setNNZ( int nnz ) { nNonZeros = nnz; }
};

int main() {
  MatrixTest mat1;

  // now set up mat1
  mat1.setNRows( nvtxs );
  mat1.setNCols( nvtxs );
  mat1.setNNZ( Aptr[nvtxs] );
  mat1.getColPtr().borrow( Aptr, nvtxs+1 );
  mat1.getRowIdx().borrow( Aind, Aptr[nvtxs] );

  // finally get mat1 to call setDiagPtr
  if ( !mat1.setDiagPtr() ) {
    cout << "mat1.setDiagPtr() failed " << endl;
  }

  // const int * colPtr = mat1.getColPtr().lend(); //unused
  const int * rowIdx = mat1.getRowIdx().lend();
  const int * diagPtr = mat1.getDiagPtr().lend();

  for( int i=0; i<nvtxs; ++i ) {
    cout << " DiagPtr[ " << i << "] =  " << rowIdx[ diagPtr[ i ] ] << endl;
  }
}
    
