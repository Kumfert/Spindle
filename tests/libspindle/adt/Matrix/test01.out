 DiagPtr[ 0] =  0
 DiagPtr[ 1] =  1
 DiagPtr[ 2] =  2
 DiagPtr[ 3] =  3
 DiagPtr[ 4] =  4
 DiagPtr[ 5] =  5
 DiagPtr[ 6] =  6
 DiagPtr[ 7] =  7
 DiagPtr[ 8] =  8
